src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/RoundaboutMissingTagCheck.java

```diff
@@ -0,0 +1,196 @@
+     */
+    private PolyLine buildOriginalOsmWayGeometry(final Edge edge)
+    {
+        // Identify all sections of original OSM way
+        final Set<Edge> edgesFormingOSMWay = new OsmWayWalker(edge).collectEdges().stream()
+                .filter(Edge::isMasterEdge).collect(Collectors.toSet());
+
+        // Sort sectioned Edges by Ids
+        final List<Edge> sortedEdges = new ArrayList<>(edgesFormingOSMWay);
+        sortedEdges.sort(Comparator.comparingLong(AtlasObject::getIdentifier));
+
+        // Build original OSM polyline
+        PolyLine geometry = new PolyLine(sortedEdges.get(0).getRawGeometry());
+        for (int i = 1; i < sortedEdges.size(); i++)
+        {
+            geometry = geometry.append(sortedEdges.get(i).asPolyLine());
+        }
+        return geometry;
+    }
+
+    /**
+     * Check if original OSM Way is intersecting with CAR_NAVIGABLE_HIGHWAYS. See {@link HighwayTag}
+     *
+     * @param edge
+     *            entity to check
+     * @return true if way intersecting with more more than
+     *         {@link RoundaboutMissingTagCheck#MINIMUM_INTERSECTION}
+     */
+    private boolean intersectingWithMoreThan(final Edge edge)
+    {
+        // Identify all sections of original OSM way
+        final Set<Edge> edgesFormingOSMWay = new OsmWayWalker(edge).collectEdges().stream()
+                .filter(Edge::isMasterEdge).collect(Collectors.toSet());
+        final Set<Long> connectedEdges = new HashSet<>();
+        edgesFormingOSMWay.forEach(obj -> obj.connectedEdges().stream().filter(Edge::isMasterEdge)
+                // intersection with navigable roads
+                .filter(HighwayTag::isCarNavigableHighway)
+                .filter(e -> e.getTag(ServiceTag.KEY).isEmpty())
+                // de-duplication sectioned edges
+                .forEach(wayId -> connectedEdges.add(wayId.getOsmIdentifier())));
+        return connectedEdges.size() > MINIMUM_INTERSECTION;
+    }
+
+    /**
+     * Check if Edge is part of Closed Way. See https://wiki.openstreetmap.org/wiki/Item:Q4669
+     *
+     * @param edge
+     *            entity to check
+     * @return true if edge is part of closed way.
+     */
+    private boolean isPartOfClosedWay(final Edge edge)
+    {
+        final HashSet<Long> wayIds = new HashSet<>();
+        Edge nextEdge = edge;
+        // Loop through out going edges with the same OSM id
+        while (nextEdge != null)
+        {
+            wayIds.add(nextEdge.getIdentifier());
+            final List<Edge> nextEdgeList = Iterables.stream(nextEdge.outEdges())
+                    .filter(Edge::isMasterEdge)
+                    .filter(outEdge -> outEdge.getOsmIdentifier() == edge.getOsmIdentifier())
+                    .collectToList();
+            nextEdge = nextEdgeList.isEmpty() ? null : nextEdgeList.get(0);
+            // If original edge is found, the way is closed
+            if (nextEdge != null && wayIds.contains(nextEdge.getIdentifier()))
```

Comment:
Because you know you are dealing with the first edge section, you could just check that the in Edges contain an edge with the same OSM id. That would save you having to traverse along the whole Way.

Reply:
Thanks for suggestion!

Reply:
In this case every way will be closedWay. Instead of "contain" we need:
return Edges.get(0).start().getLocation().equals(Edges.get(Edges.size()-1).end().getLocation());

it slower than Edge Walker.

Reply:
What about using `edge.inEdges().stream().filter(e -> e.getOsmIdentifier() == edge.getOsmIdentifier()).count() == 1`?

Reply:
slick suggestion! need to add our favorite .filter(Edge::isMasterEdge) to make it complete. thanks a lot!!